### ISSUE TITLE
Read about ajax queries

### ISSUE THREAD

**Issue by geniashand** [TechnionYP5777/Info-Evaluation#150]
As a continuation to #149 in order to execute our queries using the GUI screens there is a need to read about ajax queries

**Comment by geniashand** [comment]
AJAX is an acronym standing for Asynchronous JavaScript and XML and this technology help us to load data from the server without a browser page refresh.
First I'm going to read about [AJAX in genera](https://www.tutorialspoint.com/ajax/index.htm)l, than I'm going to move to the [jQuery tutorial](https://www.tutorialspoint.com/jquery/jquery-ajax.htm)

AJAX is a web browser technology independent of web server software.

With AJAX, when you hit submit, JavaScript will make a request to the server, interpret the results, and update the current screen. In the purest sense, the user would never know that anything was even transmitted to the server.

A user can continue to use the application while the client program requests information from the server in the background.
Intuitive and natural user interaction. Clicking is not required, mouse movement is a sufficient event trigger.

**Comment by geniashand** [comment]
AJAX cannot work independently. It is used in combination with other technologies to create interactive webpages.

- **JavaScript** - JavaScript function is called when an event occurs in a page
- **DOM** - API for accessing and manipulating structured documents.
- **CSS**
- **XMLHttpRequest** - JavaScript object that performs asynchronous interaction with the server.

**Comment by geniashand** [comment]
With AJAX we can implement a feature like google's "google suggest".

**Comment by geniashand** [comment]
The Simplest way to make your source code compatible with a browser is to use try...catch blocks in your JavaScript.
```html
<html>
<body>
   <script language="javascript" type="text/javascript">
   <!-- 
   //Browser Support Code
   function ajaxFunction(){
      var ajaxRequest;  // The variable that makes Ajax possible!

      try{
         // Opera 8.0+, Firefox, Safari 
         ajaxRequest = new XMLHttpRequest();
      }catch (e){

         // Internet Explorer Browsers
         try{
            ajaxRequest = new ActiveXObject("Msxml2.XMLHTTP");
         }catch (e) {
            try{
               ajaxRequest = new ActiveXObject("Microsoft.XMLHTTP");
            }catch (e){

               // Something went wrong
               alert("Your browser broke!");
               return false;
            }
         }
      }
   }
   //-->
   </script>
   
   <form name='myForm'>
      Name: <input type='text' name='username' /> <br />
      Time: <input type='text' name='time' />
   </form>
   
</body>
</html>
```

In the above JavaScript code, we try three times to make our XMLHttpRequest object. Our first attempt:
`ajaxRequest = new XMLHttpRequest();`
It is for Opera 8.0+, Firefox, and Safari browsers. If it fails, we try two more times to make the correct object for an Internet Explorer browser with:
`ajaxRequest = new ActiveXObject("Msxml2.XMLHTTP");`
`ajaxRequest = new ActiveXObject("Microsoft.XMLHTTP");`
If it doesn't work, then we can use a very outdated browser that doesn't support XMLHttpRequest, which also means it doesn't support Ajax.

Most likely though, our variable ajaxRequest will now be set to whatever XMLHttpRequest standard the browser uses and we can start sending data to the server.

**Comment by geniashand** [comment]
## Steps of AJAX Operation
- A client event occurs.
- An XMLHttpRequest object is created.
- The XMLHttpRequest object is configured.
- The XMLHttpRequest object makes an asynchronous request to the Webserver.
- The Webserver returns the result containing XML document.
- The XMLHttpRequest object calls the callback() function and processes the result.
- The HTML DOM is updated.

[gread read](https://www.tutorialspoint.com/ajax/ajax_in_action.htm) on how to set up an AJAX event

**Comment by geniashand** [comment]
## XMLHttpRequest
XMLHttpRequest (XHR) is an API that can be used by JavaScript, JScript, VBScript, and other web browser scripting languages to transfer and manipulate XML data to and from a webserver using HTTP, establishing an independent connection channel between a webpage's Client-Side and Server-Side.
XMLHttpRequest can be used to fetch data in other formats, e.g. JSON or even plain text.

### XMLHttpRequest Methods
- **abort()**
    Cancels the current request.
- **getAllResponseHeaders()**
    Returns the complete set of HTTP headers as a string.
- **getResponseHeader( headerName )**
    Returns the value of the specified HTTP header.
- **open( method, URL )
open( method, URL, async )
open( method, URL, async, userName )
open( method, URL, async, userName, password )**
Specifies the method, URL, and other optional attributes of a request.
(method parameter can be: "GET", "POST", or "HEAD". If we choose REST we can use "PUT" and "DELETE" too)
The "async" parameter specifies whether the request should be handled asynchronously or not. "true" means that the script processing carries on after the send() method without waiting for a response, and "false" means that the script waits for a response before continuing script processing.
- **send( content )**
   Sends the request.
- **setRequestHeader( label, value )**
   Adds a label/value pair to the HTTP header to be sent.

You can read [here](https://www.tutorialspoint.com/ajax/what_is_xmlhttprequest.htm) about the properties XMLHttpRequest has

**Comment by geniashand** [comment]
**AJAX can work with SQL (which is great for us)**

Lets assume we have the table below

Name | Age | Sex | Wpm     
--- | --- | --- | ---
Jerry  | 120 | m | 20

The client side HTML file

``` html
<html>
<body>
<script language="javascript" type="text/javascript">
<!-- 
//Browser Support Code
function ajaxFunction(){
   var ajaxRequest;  // The variable that makes Ajax possible!
   try{
   
      // Opera 8.0+, Firefox, Safari
      ajaxRequest = new XMLHttpRequest();
   }catch (e){
      
      // Internet Explorer Browsers
      try{
         ajaxRequest = new ActiveXObject("Msxml2.XMLHTTP");
      }catch (e) {
         
         try{
            ajaxRequest = new ActiveXObject("Microsoft.XMLHTTP");
         }catch (e){
         
            // Something went wrong
            alert("Your browser broke!");
            return false;
         }
      }
   }
   
   // Create a function that will receive data
   // sent from the server and will update
   // div section in the same page.
   ajaxRequest.onreadystatechange = function(){
   
      if(ajaxRequest.readyState == 4){
         var ajaxDisplay = document.getElementById('ajaxDiv');
         ajaxDisplay.innerHTML = ajaxRequest.responseText;
      }
   }
   
   // Now get the value from user and pass it to
   // server script.
   var age = document.getElementById('age').value;
   var wpm = document.getElementById('wpm').value;
   var sex = document.getElementById('sex').value;
   var queryString = "?age=" + age ;
   
   queryString +=  "&wpm=" + wpm + "&sex=" + sex;
   ajaxRequest.open("GET", "ajax-example.php" + queryString, true);
   ajaxRequest.send(null); 
}
//-->
</script>

<form name='myForm'>

   Max Age: <input type='text' id='age' /> <br />
   Max WPM: <input type='text' id='wpm' /> <br />
   Sex: 
   <select id='sex'>
      <option value="m">m</option>
      <option value="f">f</option>
   </select>
   <input type='button' onclick='ajaxFunction()' value='Query MySQL'/>
   
</form>
<div id='ajaxDiv'>Your result will display here</div>
</body>
</html>
```
